### ISSUE TITLE
Server configured to use TLS accepts logs from any client with a self-signed cert

### ISSUE THREAD

**Issue by rtkpmcalpine** [rsyslog/librelp#48]
While evaluating librelp I've discovered a bug where any client is able to send a server messages, even though the server is configured to validate the client's certificate and has permittedpeer set.

My desired configuration is to have both client and server validate each others' certificates, and validate their CN against the names in the tls.permittedpeer list.

The behaviour I observed is:
- the client validates the server's cert and validates the CN matches permittedpeer (this is good.)
- If the client has a cert which passes CA validation and the CN matched permittedpeer then messages are accepted (this is good)
- If the client has a cert which passes CA validation and the CN does NOT match permittedpeer then messages are NOT accepted (this is good)
- If the client has a self-signed cert which does not pass CA validation then messages are accepted (this is BAD!)

I would expect that if the client's cert does not pass validation then the connection should be closed immediately.

Configuration for the server: [rsyslog-server.txt](https://github.com/rsyslog/librelp/files/382355/rsyslog-server.txt)

Configuration for the client: [rsyslog-client.txt](https://github.com/rsyslog/librelp/files/382356/rsyslog-client.txt)

Commands used to generate CA and certs: [genkeys.txt](https://github.com/rsyslog/librelp/files/382357/genkeys.txt)

Command to generate logs: date | nc -vv 127.0.0.1 5514
- With the configurations as-is, the client will connect to the server successfully. (this is good)
- Changing the permittedpeer option to something else on either client or server results in the connection failing (this is good)
- In the server's config: Changing server.key/cert to fake.key/cert results in the client rejecting the connection (this is good)
- In the client's config: Changing client.key/cert to fake.key/cert results in the server accepting the connection (this is BAD)

**Comment by rgerhards** [comment]
@PascalWithopf could you see if you can reproduce this?

**Comment by PascalWithopf** [comment]
I tried to reproduce this, but it is working fine.
Is there anything else I should to know about your configuration?
@rtkpmcalpine could you provide me with your debug logs, especially from the server.

**Comment by kaizera071** [comment]
Hi there i had the same issue, and I find out that the problem is not in librelp.
In my case the problem was in gnutls (3.0.11 stock gnutls version in ubuntu 14.04 trusty), and after update with the newest gnutls version the problem disappeared.

**Comment by voelzmo** [comment]
Hey @rgerhards and @PascalWithopf I'm not sure why this issue hasn't seen much attention so far, but the initial bug report sounds quite disturbing to me. We have two different configurations, one of them shows the above issue, while the other doesn't. Could you please have a more detailed look why this might be the case?

Configuration 1, showing the above issue
```
* Ubuntu Trusty LTS
* rsyslog 8.30.0
* librelp 1.2.14
* libgnutls26 2.12.23-12ubuntu2.6
```

Configuration 2, *not* showing the above issue
```
* Ubuntu Xenial
* rsyslog 8.30.0
* librelp 1.2.14
* libgnutls30 3.4.10-4ubuntu1.3
```

I'm not sure if this really is an issue with gnutls26, given that there are no CVEs open for it in the [Ubuntu USNs](https://people.canonical.com/~ubuntu-security/cve/pkg/gnutls26.html).  It could be that something changed in the API between gnutls26 and gnutls30?

**Comment by rgerhards** [comment]
@voelzmo The problem is that we cannot reproduce it... Could you contribute a debug log and build rsyslgo from source? If so, we could go through a couple of iterations trying to nail it in your environment.

**Comment by landzhev** [comment]
Hi @rgerhards and @PascalWithopf,

We are using rsyslog in our solution and we have reproduced the issue in our test landscapes.
It would be great if we have an online meeting and try to reproduce it together in your test environment.
We also considered adding additional debug logs, but that would cost a lot of effort as our C expertise is small.
Can we schedule a meeting? We can do it at whatever time is convenient for you?

Best regards,
Nikolay

**Comment by rgerhards** [comment]
@alorbach can you verify (best with a test?) that this does at least not happen if build with openssl support?

**Comment by rgerhards** [comment]
fixed via #111
